### PR TITLE
Remove tokio runtime, use Context<'_>'s Waker for networking

### PR DESCRIPTION
**Pull request by MingweiSamuel**

### CODE REVIEW

File: hydroflow/src/scheduled/mod.rs [hunk 1]
```diff
@@ -139,6 +135,18 @@ impl Hydroflow {
         }
     }
 
+    /**
+     * Run the dataflow graph to completition asynchronously.
+     */
+    pub async fn run_async(&mut self) -> Result<!, RecvError> {
+        loop {
+            // println!("HELLO WORLD {}", self.ready_queue.len());
```
Comment (justinj) [inline hydroflow/src/scheduled/mod.rs:143]: nit: stray comment

File: hydroflow/src/scheduled/net/mod.rs [hunk 2]
```diff
@@ -75,48 +59,58 @@ impl Hydroflow {
             reader.map(|buf| Some(<Message>::decode(&buf.unwrap().into()))),
         );
         let mut writer = FramedWrite::new(writer, LengthDelimitedCodec::new());
-        let rt = self.rt.clone();
+
+        let mut queue = VecDeque::new();
```
Comment (justinj) [inline hydroflow/src/scheduled/net/mod.rs:63]: nit: I think `queue` here could use a slightly more descriptive name. `message_queue`?

**Review by justinj — APPROVED** [review]
Looks good to me! Super glad to see all those `block_on`s go away, and I'm also very happy that it's still possible to run Hydroflow synchronously.

A couple questions:
1. What does yielding semantically here? It hops up _completely_ out of the Hydroflow runtime and into the Tokio runtime, correct? Is there any world where we'd want to leverage Tokio yielding within an operator to another operator or is that muddling some concerns?
2.  I don't fully understand why we had to delete the `Net` trait? I don't really mind but I think it would be really nice if the networking stuff could be in a different crate entirely—is that possible with this setup? I don't see why it wouldn't be.

File: hydroflow/src/scheduled/mod.rs (hunk 1 above)
Comment (MingweiSamuel) [reply hydroflow/src/scheduled/mod.rs:143]: Thx

**Comment by MingweiSamuel** [comment]
>     1. What does yielding semantically here? It hops up _completely_ out of the Hydroflow runtime and into the Tokio runtime, correct? Is there any world where we'd want to leverage Tokio yielding within an operator to another operator or is that muddling some concerns?

Yes, it basically says "yield me immediately but reschedule me to run again later" at the Tokio runtime level. The `Hydroflow` graph acts as its own mini-runtime hiding within the Tokio runtime (if you use HF's async features).

[The code it runs](https://docs.rs/tokio/1.14.0/src/tokio/task/yield_now.rs.html#38-59) is actually not magic at all (and not Tokio-specific). You could do the equivalent thing in a subgraph with `cx.waker().wake_by_ref();` and then returning, which would reschedule the subgraph after the rest of the pending subgraphs have run (in the current scheduler). But it's not possible to access the Tokio runtime from within the subgraphs.

>     2. I don't fully understand why we had to delete the `Net` trait? I don't really mind but I think it would be really nice if the networking stuff could be in a different crate entirely—is that possible with this setup? I don't see why it wouldn't be.

The short answer is that I made the functions async (rather than having a hidden runtime inside them), and to write async traits with lifetimes you need GATs. Which is actually reasonably possible to do, but messy. There are a lot of alternative ways to structure a regular trait though. I'll probably revisit it in the next few PRs

**Pull request merged**